server/resource_server/service/service.go

```diff
@@ -277,6 +272,25 @@ func (s *service) RegisterUserPrecheck(req dto.RegisterUserPrecheckDTO, iterCoun
 	return rmSalt, nil
 }
 
+func (s *service) UpdateUserPassword(username string, newPassword string, salt string) error {
+	hashedPassword := utils.SaltAndHashPassword(newPassword, salt)
+	return s.repository.UpdateUserPassword(username, hashedPassword)
+}
+
+func (s *service) UpdateUserPasswordV2(username string, storedKey string, serverKey string) error {
+	if username == "" {
+		return fmt.Errorf("invalid username")
+	}
+	if storedKey == "" {
+		return fmt.Errorf("invalid stored key")
+	}
+	if serverKey == "" {
+		return fmt.Errorf("invalid server key")
+	}
```

Comment:
It might be better to add those checks into the dto structure directly, similar to how it's done in `RegisterUserDTO`:
```Username    string `json:"username" validate:"required,min=3,max=50"` ```
In this case, you could remove them from service.go as well.